isabelle/Doc/Specification/Guarantees.thy

```diff
@@ -14,132 +14,191 @@ begin
+in the MList.thy theory\<close>
 
 text \<open>
+\<^item> valid\_map :: @{typeof valid_map}
+\<close>
 
-One of the dangers of using smart contracts is that a badly written one can potentially lock its
-funds forever.\<close>
-text \<open>By the end of a Marlowe contract, all the assets paid to the contract must be returned to a subset of the participants. To ensure this behaviour we
-proved two properties: ``MultiAssets Preservation'' and ``Contracts Always Close''.\<close>
+text \<open>Moreover, we use the following function to ensure that the Maps are valid and the
+ \<^emph>\<open>accounts\<close> only have positive values\<close>
 
 text \<open>
-Regarding asset preservation, tokens are not created nor destroyed by the semantics. Formally speaking,
-if a transaction is computed successfully, the sum of the assets stored in the previous state and
-the assets deposited by a transaction are equal to the assets in the new state and possible external payments
-(from internal account to external party).
+\<^item> validAndPositive\_state :: @{typeof validAndPositive_state}
 \<close>
 
-text \<open>@{thm computeTransaction_preserves_assets}\<close>
+text \<open>
+The following theorem asserts that if the State is valid before computing a transaction,
+it is also valid after.
+\<close>
 
-text \<open>If calling computeTransaction yields an error, then the transaction is invalid and should not be
-part of the blockchain, preserving assets in its current state.\<close>
+text \<open>\<^bold>\<open>lemma\<close> \<^emph>\<open>computeTransaction\_preserves\_validAndPositive\_state\<close>:@{thm [display,names_short, margin=40] computeTransaction_preserves_validAndPositive_state}\<close>
+
+text \<open>The function and lemma are located in \<^emph>\<open>PositiveAccounts.thy\<close>.\<close>
 
-section \<open>Contracts Always Close\<close>
+section \<open>Finite contracts\label{sec:finite-contracts}\<close>
 
 text \<open>
+All contracts are finite and have an expiration time. We can use the following function
+(located in Semantics.thy) to see what that time is\<close>
 
-For every Marlowe Contract there is a maximum time that allows any participant
-to create an empty transaction that will close the contract successfully.
+(* We use a hardcoded type as aliases get expanded *)
+text \<open>
+\<^item> maxTimeContract :: \isa{Contract\ {\isasymRightarrow}\ POSIXTime}\<close>
+
+text 
+\<open>
+Contracts in most blockchains (including Cardano) need to be triggered by a transaction in order
+to progress. Because of this, they can't automatically refund participants
+after the contract timeouts on its own. Instead, after it has expired, the contract will not accept any
+more inputs, it will only accept a transaction with no inputs, and that transaction will trigger the refunds
+if any and close the contract. To ensure this, we prove that a transaction with no inputs with a
+validity interval that starts after the max time will always succeed
```

Comment:
```suggestion
after the contract times out on its own. Instead, after it has expired, the contract will not accept any
more inputs, it will only accept a transaction with no inputs, and that transaction will trigger the refunds,
if any, and close the contract. To ensure this, we prove that a transaction with no inputs and a
validity interval that starts after the max time will always succeed
```

Reply:
I should review my own suggestions